### PR TITLE
allowJs:true + declaration:true

### LINKED ISSUE #7546: Allow `--declaration` with `--allowJs`
From https://github.com/Microsoft/TypeScript/issues/7535, it would be useful for JavaScript package authors to have the ability to generate declarations from their project. This would help with getting started on a type definition (it would be populated with known interfaces/functions/types that are exported) and would also allow the creation of continuous integration scripts that check the JavaScript types with the hand-coded definition for inconsistencies.

### PR DESCRIPTION
**Pull request by nojvek**
- [x] There is an associated issue. Fixes #7546
 - [x] Code is up-to-date with the `master` branch
 - [x] You've successfully run `jake runtests` locally
 - [x] You've signed the CLA
 - [x] There are new or updated unit tests validating the change

** TODO **

From previous PR: #15911

 - [ ] Fix commented out tsserver test. Not sure what the correct fix is 
 - [ ] ES5 prototype style function definitions
```
function f() {
}
f.prototype.method = function() {}   // should probably emit a class?
```
 - [ ] @typedef {Object} SpecialType -> Interfaces

```
/**
 * @typedef {object} SpecialType1 - creates a new type named 'SpecialType'
 * @property {string} prop1 - a string property of SpecialType
 * @property {number} prop2 - a number property of SpecialType
 * @property {number=} prop3 - an optional number property of SpecialType
 */
/** @type {SpecialType1} */
var specialTypeObject1;
```

 - [ ] Recognize @private/@readonly -> private X / readonly X

```
  /** @private */
  _render() {...}

  /** @readonly */
  static observedAttributes() {...}
```

### CODE REVIEW

File: src/harness/unittests/tscWatchMode.ts [hunk 1]
```diff
@@ -1016,46 +1016,46 @@ namespace ts.tscWatch {
...
-                path: "/a/b/app.ts",
-                content: "let x = 10"
-            };
-            const configFileContentBeforeComment = `{`;
-            const configFileContentComment = `
-                    // comment
-                    // More comment`;
-            const configFileContentAfterComment = `
-                    "compilerOptions": {
-                        "allowJs": true,
```
Comment (sheetalkamat) [inline src/harness/unittests/tscWatchMode.ts]: you could change this to set { "inlineSourceMap": true, "sourceMap": true } to make sure the errors are reported on those lines in the config file.

**Comment by TeffenEllis** [comment]
This looks awesome! Is there an ETA to release?

**Comment by nojvek** [comment]
Needs some work on emitting part for type interfaces described in jsdocs.
I’ve been using puppeteer project as a real world test case to solve.

Typescript doesn’t understand that module.exports should also export the
interfaces so need to solve that too.

I think a lot of these can be sped up if I could get a mentor from
Typescript team.

This isn’t as easy as I thought it would be.

Please help Typescript.

**Comment by DSchroer** [comment]
I did a quick test of this on a project I am developing for. It works great except that it automatically detects `require("SOMETHING")` as a module import even if the TypeScript module setting is set to none.

This is a weird quirk of the code I am dealing with but it would be nice to be able to disable module detection. Since in my case require is just another function.

**Comment by nojvek** [comment]
Yeah there are lots of little quirks like that to work out. 

But being able to auto-generate types for js projects is going to be amazing.

**Comment by weswigham** [comment]
@nojvek Once #21930 is merged, I'd like to write a transformer for JS files that simply iterates over their exported symbols, gets their declared types, and uses a modified `typeToTypeNode` on them to generate declarations for all required symbols. It needs to be a very different process from our usual subtractive declaration emit.

**Comment by nojvek** [comment]
ooo I like it. Do you have a guesstimate for the merge ?

**Comment by ahocevar** [comment]
What needs to be done to finish this effort?

**Comment by nojvek** [comment]
I basically just need to re-focus and get back into it. TSC has moved some pieces around so need to re-base and figure out the merge conflicts + tests.

Hopefully we don't end up bikeshedding again. Been trying to land this for a couple of years now :/

**Comment by weswigham** [comment]
You _probably_ want to start fresh rather than attempt a rebase - the declaration emit pipeline has entirely changed since this PR was last updated.

**Comment by RyanCavanaugh** [comment]
@nojvek thanks for the work so far! I'm going to close this since it's going to require a substantial rebasing anyway. Looking forward to the next iteration.

**Pull request closed without merging**